### PR TITLE
Refactor TransactionBuilder

### PR DESCRIPTION
**Pull request by NicolasDorier**
Refactoring transaction builder. In some corner case situation, the transaction builder can pay less than the relay fee. Fixing this bug is not easy, so I need to refactor step by step this spaghetti.

### CODE REVIEW

File: NBitcoin/Utils.cs [hunk 1]
```diff
@@ -739,21 +739,25 @@ public static void Shuffle<T>(T[] arr, Random rand)
...
 			rand = rand ?? new Random();
-			for (int i = 0; i < arr.Count; i++)
+			for (int i = start; i < arr.Count; i++)
 			{
-				var fromIndex = rand.Next(arr.Count);
+				var fromIndex = rand.Next(start, arr.Count - start);
 				var from = arr[fromIndex];
 
-				var toIndex = rand.Next(arr.Count);
+				var toIndex = rand.Next(start, arr.Count - start);
```
Comment (lontivero) [inline NBitcoin/Utils.cs:750]: This works okay but you could use some syntax sugar, something from:
 
```c#
var segment = arr[start..]; 

var fromIdx = rand.TakeOne(segment); 
var toIdx = rand.TakeOne(segment); 
(segment[fromIdx], segment[toIdx]) = (segment[toIdx], segment[fromIdx]);
```

Comment (NicolasDorier) [reply NBitcoin/Utils.cs:750]: I must admit that I have hard time to understand what it does :p

Comment (dangershony) [reply NBitcoin/Utils.cs:750]: ha I agree I am not sure I would call this sugar `arr[start..]`

**Comment by NicolasDorier** [comment]
The old TransactionBuilder had a slight change to generate a transaction with fee lower than the feerate if the coinselector was chosing a different set of coins than it did when `SendEstimatedFee` was called.

This was a very rare bug, actually found it via spurious tests running the builder thousands of time as the CI was sometimes failing the test.

The refactoring took me 3 fucking days. This is the trickiest code of NBitcoin. But now it is not only more readable, with the bug fixed, but also it will generate transactions with way closer to the expected fee rate!

**Pull request merged**